File: app/src/main/java/com/example/ldwa/DataAdapter.java
package com.example.ldwa;

import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class DataAdapter {

    protected static final String TAG = "DataAdapter";

    private final Context mContext;
    private SQLiteDatabase mDb;
    private DataBaseHelper mDbHelper;

    public DataAdapter(Context context) {
        this.mContext = context;
        mDbHelper = new DataBaseHelper(mContext);
    }

    public DataAdapter createDatabase() throws SQLException {
        try {
            mDbHelper.createDataBase();
        } catch (IOException mIOException) {
            Log.e(TAG, mIOException.toString() + "  UnableToCreateDatabase");
            throw new Error("UnableToCreateDatabase");
        }
        return this;
    }

    public DataAdapter open() throws SQLException {
        try {
            mDbHelper.openDataBase();
            mDbHelper.close();
            mDb = mDbHelper.getReadableDatabase();
        } catch (SQLException mSQLException) {
            Log.e(TAG, "open >>"+ mSQLException.toString());
            throw mSQLException;
        }
        return this;
    }

    public void close() {
        mDbHelper.close();
    }

    public Cursor selectAll() {
        try {
            Cursor mCur = mDb.rawQuery("select * from weapons order by name", null);
            if (mCur != null) {
                mCur.moveToNext();
            }
            return mCur;
        } catch (SQLException mSQLException) {
            Log.e(TAG, "getTestData >>"+ mSQLException.toString());
            throw mSQLException;
        }
    }


    private static class DataBaseHelper extends SQLiteOpenHelper {

        private static String TAG = "DataBaseHelper"; // Tag just for the LogCat window
        private static String DB_NAME ="weapons_database.sqlite"; // Database name
        private static int DB_VERSION = 1; // Database version
        private final File DB_FILE;
        private SQLiteDatabase mDataBase;
        private final Context mContext;

        public DataBaseHelper(Context context) {
            super(context, DB_NAME, null, DB_VERSION);
            DB_FILE = context.getDatabasePath(DB_NAME);
            this.mContext = context;
        }

        public void createDataBase() throws IOException {
            // If the database does not exist, copy it from the assets.
            boolean mDataBaseExist = checkDataBase();
            if(!mDataBaseExist) {
                this.getReadableDatabase();
                this.close();

                // Copy the database from assests
                copyDataBase();
                Log.e(TAG, "createDatabase database created");

            }
        }

        // Check that the database file exists in databases folder
        private boolean checkDataBase() {
            return DB_FILE.exists();
        }

        // Copy the database from assets
        private void copyDataBase() throws IOException {
            InputStream mInput = mContext.getAssets().open(DB_NAME);
            OutputStream mOutput = new FileOutputStream(DB_FILE);
            byte[] mBuffer = new byte[1024];
            int mLength;
            while ((mLength = mInput.read(mBuffer)) > 0) {
                mOutput.write(mBuffer, 0, mLength);
            }
            mOutput.flush();
            mOutput.close();
            mInput.close();
        }

        // Open the database, so we can query it
        public boolean openDataBase() throws SQLException {
            // Log.v("DB_PATH", DB_FILE.getAbsolutePath());
            // mDataBase = SQLiteDatabase.openDatabase(DB_FILE, SQLiteDatabase.CREATE_IF_NECESSARY);
            mDataBase = SQLiteDatabase.openDatabase(String.valueOf(DB_FILE), null, SQLiteDatabase.NO_LOCALIZED_COLLATORS);
            return mDataBase != null;
        }

        @Override
        public synchronized void close() {
            if(mDataBase != null) {
                mDataBase.close();
            }
            super.close();
        }

        @Override
        public void onCreate(SQLiteDatabase db) {

        }

        @Override
        public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {

        }

    }
}

File: android/app/src/main/java/com/danielschlatter/ldwa/DataAdapter.kt
package com.danielschlatter.ldwa

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

open class DataAdapter(context: Context) {
    private var mDb: SQLiteDatabase? = null
    private val mDbHelper: DataBaseHelper = DataBaseHelper(context)

    @Throws(SQLException::class)
    fun createDatabase() {
        try {
            mDbHelper.createDataBase()
        } catch (mIOException: IOException) {
            Log.e(TAG, "$mIOException  UnableToCreateDatabase")
            throw Error("UnableToCreateDatabase")
        }
    }

    @Throws(SQLException::class)
    fun open() {
        mDb = try {
            mDbHelper.openDataBase()
            mDbHelper.close()
            mDbHelper.readableDatabase
        } catch (mSQLException: SQLException) {
            Log.e(TAG, "open >>$mSQLException")
            throw mSQLException
        }
    }

    fun selectAll(): Cursor {
        return try {
            mDb!!.rawQuery("select * from weapons order by name", null)
        } catch (mSQLException: SQLException) {
            Log.e(TAG, "selectall failed $mSQLException")
            throw mSQLException
        }
    }

    fun selectPerk(name: String): String {
        return try {
            val mCur = mDb!!.rawQuery(
                "select displayProperties_icon from perks where displayProperties_name = ?",
                arrayOf(name)
            )
            mCur.moveToNext()
            val path = mCur.getString(mCur.getColumnIndex("displayProperties_icon"))
            mCur.close()
            path.substring(25, path.length - 4).replace('/', '_')
        } catch (mSQLException: SQLException) {
            Log.e(TAG, "selectperk failed$mSQLException$name")
            throw mSQLException
        }
    }

    fun insertRoll(weapon: String?, p1: Int, p2: Int, p3: Int, p4: Int) {
        val cv = ContentValues()
        cv.put("weapon", weapon)
        cv.put("p1", p1)
        cv.put("p2", p2)
        cv.put("p3", p3)
        cv.put("p4", p4)
        mDb!!.insert("saved_rolls", null, cv)
    }

    fun removeRoll(weapon: String, p1: Int, p2: Int, p3: Int, p4: Int) {
        mDb!!.delete(
            "saved_rolls",
            "weapon = ? and p1 = ? and p2 = ? and p3 = ? and p4 = ?",
            arrayOf(weapon, p1.toString(), p2.toString(), p3.toString(), p4.toString())
        )
    }

    fun selectRolls(weapon: String): Cursor {
        return mDb!!.rawQuery("select * from saved_rolls where weapon = ?", arrayOf(weapon))
    }

    private class DataBaseHelper(context: Context) :
        SQLiteOpenHelper(context, DB_NAME, null, DB_VERSION) {
        private val dbFile: File = context.getDatabasePath(DB_NAME)
        private var mDataBase: SQLiteDatabase? = null
        private val mContext: Context = context

        @Throws(IOException::class)
        fun createDataBase() {
            // If the database does not exist, copy it from the assets.
            val mDataBaseExist = checkDataBase()
            if (!mDataBaseExist) {
                this.readableDatabase
                close()

                // Copy the database from assets
                copyDataBase()
                Log.e(TAG, "createDatabase database created")
                openDataBase()
                try {
                    mDataBase!!.execSQL("create table saved_rolls(weapon text, p1 int, p2 int, p3 int, p4 int)")
                } catch (mSQLException: SQLException) {
                    Log.e(TAG, "table select_rolls creation failed >>$mSQLException")
                    throw mSQLException
                }
            }
        }

        // Check that the database file exists in databases folder
        private fun checkDataBase(): Boolean {
            return dbFile.exists()
        }

        // Copy the database from assets
        @Throws(IOException::class)
        private fun copyDataBase() {
            val mInput = mContext.assets.open(DB_NAME)
            val mOutput: OutputStream = FileOutputStream(dbFile)
            val mBuffer = ByteArray(1024)
            var mLength: Int
            while (mInput.read(mBuffer).also { mLength = it } > 0) {
                mOutput.write(mBuffer, 0, mLength)
            }
            mOutput.flush()
            mOutput.close()
            mInput.close()
        }

        // Open the database, so we can query it
        @Throws(SQLException::class)
        fun openDataBase() {
            mDataBase = SQLiteDatabase.openDatabase(
                dbFile.toString(),
                null,
                SQLiteDatabase.NO_LOCALIZED_COLLATORS
            )
        }

        @Synchronized
        override fun close() {
            if (mDataBase != null) {
                mDataBase!!.close()
            }
            super.close()
        }

        override fun onCreate(db: SQLiteDatabase) {}
        override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {}

        companion object {
            private const val TAG = "DataBaseHelper" // Tag just for the LogCat window
            private const val DB_NAME = "weapons_database.sqlite" // Database name
            private const val DB_VERSION = 1 // Database version
        }

    }

    companion object {
        protected const val TAG = "DataAdapter"
    }

}